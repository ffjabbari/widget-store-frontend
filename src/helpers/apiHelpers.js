/** API Helper functions. */

// local imports
import { asyncAPIRequest } from '../hooks/apiHook';

/**
 * updateUserProfile()
 *
 * Make patch request to update a user profile. Used by CustomerAddressForm
 * to allow updating customer address of signed in users.
 *
 * @param {object} values - form values
 * @param {object} user - user object
 * @param {token} token - user token
 * @param {function} setResponseError - state setter for error display
 * @param {any} defaultAddress - Falsey or obj with address components
 * @param {function} setFormDisabled - state setter to hide form after update
 *
 * returns: true if an error happened else undefined
 */

export const updateUserProfile = async (
  values,
  user,
  token,
  setResponseError,
  defaultAddress,
  setFormDisabled
) => {
  const userUpdateData = { ...values };
  // remove email to dis-allow email updating from here
  delete userUpdateData['email'];
  delete userUpdateData['user_id'];

  // remomve key/value pairs having a value of empty string from userUpdateData
  Object.entries(userUpdateData).forEach(([key, value]) => {
    if (value === '') delete userUpdateData[key];
  });

  try {
    // make patch request
    const resp = await asyncAPIRequest(
      `/users/${user.username}`,
      'patch',
      userUpdateData,
      { _token: token }
    );
    if (resp.error) {
      setResponseError(resp.error.response.data.message);
      return true;
    } else {
      defaultAddress.current = values;
      setFormDisabled(true);
    }
  } catch (error) {
    setResponseError(error.message);
    return true;
  }
};
