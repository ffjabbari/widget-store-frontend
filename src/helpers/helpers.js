import { Redirect } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
