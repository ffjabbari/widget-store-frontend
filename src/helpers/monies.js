import Decimal from 'decimal.js';

const SF_SALES_TAX = new Decimal('0.085');

// calculate discounted price
export function calculateDiscountPrice(product) {
  if (product.discount === '0.00') return product.price;
  const discount = new Decimal(product.discount),
    price = new Decimal(product.price);
  return price.minus(price.times(discount)).toFixed(2);
}

/** Calculate discount as a percent. */
export function calculateDiscountPercent(product) {
  const discount = new Decimal(product.discount);
  return discount.times(100).toFixed(0);
}

/** Return discounted price and discount as a percent */
export function getDiscountData(product) {
  return {
    discountPrice: calculateDiscountPrice(product),
    percent: calculateDiscountPercent(product),
  };
}

/** Calculate tax if customer is in California */
export function calculateTax(subtotal, state) {
  if (!/^ca(lifornia)?$/i.test(state)) return '0.00';
  subtotal = new Decimal(subtotal);
  return subtotal.times(SF_SALES_TAX).toFixed(2);
}

/** Calculate tax plus subtotal plus shipping cost if provided */
export function calculateTotal(subtotal, state, shippingCost) {
  const tax = calculateTax(subtotal, state);
  subtotal = new Decimal(subtotal);
  shippingCost = shippingCost || 0;
  return subtotal.plus(tax).plus(shippingCost).toFixed(2);
}
