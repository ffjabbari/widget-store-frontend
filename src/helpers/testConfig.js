/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'A good remote.',
    discount: '0.15',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    rating: null,
    quantity: 1,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: '',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
    set_default: false,
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        date_added: '2021-01-24T08:00:00.000Z',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 2,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };
};
