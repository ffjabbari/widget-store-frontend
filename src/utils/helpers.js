import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
