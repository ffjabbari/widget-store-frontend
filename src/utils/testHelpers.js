import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
