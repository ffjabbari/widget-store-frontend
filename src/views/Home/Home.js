/** Home page view. */

import { useEffect, useRef, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { motion } from 'framer-motion';
import { useSelector, useDispatch } from 'react-redux';

// local imports
import Navbar from '../../components/Navbar/Navbar';
import ProductPanel from '../../components/ProductPanel/ProductPanel';
import toggleDir from '../../redux/actions/animation/toggleDir';
import StorePageRoutes from '../../routes/storePageRoutes';
import ProductList from '../../components/ProductList/ProductList';
import SearchBar from '../../components/SearchBar/SearchBar';
import pageScrollTopHook from '../../hooks/pageScrollTopHook';
import { getPathRoot } from '../../utils/helpers';
import './Home.css';

function Home() {
  const dispatch = useDispatch();
  const location = useLocation();
  const pathRoot = getPathRoot(location.pathname);
  const slideoutDir = useSelector(state => state.animation.flipFlop);
  // searchbar / productlist filtering
  const [params, setParams] = useState(null);
  const initFormState = useRef({});

  // scroll to top of page with load.
  // needed for mobile behavior.
  pageScrollTopHook();

  // toggle animation direction for page transitions.
  useEffect(() => {
    dispatch(toggleDir());
  }, [dispatch]);

  return (
    <>
      <Navbar />
      {/* Animate page transition. */}
      <motion.div
        initial={{ opacity: 0 }}
        animate={{ opacity: 1 }}
        exit={{
          x: `${slideoutDir ? '' : '-'}200vw`,
          opacity: 0,
        }}
        transition={{ duration: 0.5 }}
      >
        <ProductPanel title={'Widget Store'} byline={'We have what you need!'}>
          <SearchBar setParams={setParams} initFormState={initFormState} />
          <ProductList params={params} />
        </ProductPanel>
      </motion.div>
      {/* End animate page transition. */}

      {/* Nested Routes for product detail, cart, and login. */}
      <StorePageRoutes pathRoot={pathRoot} modalType={2} />
      {/* End Routes. */}
    </>
  );
}

export default Home;
