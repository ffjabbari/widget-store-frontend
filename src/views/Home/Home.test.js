/** Home tests */

// local imports
import { act } from '@testing-library/react';
import { renderWithStore } from '../../helpers/testHelpers';
import Home from './Home';

const axios = require('axios');
jest.mock('axios');

beforeEach(() => {
  axios.mockResolvedValue({
    data: {
      products: [
        {
          byline: 'Latest generation smart televison 2021',
          discount: '0.00',
          image_url:
            'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
          name: 'Samsung XG-900 55 inch LCD TV',
          price: '400.40',
          product_id: 1,
          rating: null,
        },
      ],
    },
  });
});

afterEach(() => {
  jest.resetAllMocks();
});

test('renders Home', async () => {
  await act(async () => {
    renderWithStore(<Home />);
  });
});

test('Home snapshot', async () => {
  let asFragment;
  await act(async () => {
    const resp = renderWithStore(<Home />);
    asFragment = resp.asFragment;
  });
  expect(asFragment()).toMatchSnapshot();
});

test('Product name in document', async () => {
  let getByText;
  await act(async () => {
    const resp = renderWithStore(<Home />);
    getByText = resp.getByText;
  });
  const cardText = getByText('Samsung XG-900 55 inch LCD TV');
  expect(cardText).toBeInTheDocument();
});
