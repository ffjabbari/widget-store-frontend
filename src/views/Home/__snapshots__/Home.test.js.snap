// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ fijioT"
  >
    <nav
      class="Navbar navbar-dark navbar navbar-expand-md navbar-light"
    >
      <span
        class="navbar-text"
      >
        <a
          class="navbar-brand brand-style"
          href="/shop"
        >
          Widget Store
        </a>
      </span>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="pr-4 navbar-collapse collapse"
      >
        <div
          class="mr-auto navbar-nav"
        >
          <span
            class="navbar-text"
          >
            <a
              href="/new"
            >
              New Products
            </a>
          </span>
          <span
            class="navbar-text"
          >
            <a
              href="/deals"
            >
              Deals
            </a>
          </span>
          <div
            class="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle nav-link"
              href="#"
              role="button"
            >
              More
            </a>
          </div>
        </div>
        <a
          href="//login"
        >
          Login
        </a>
        <span
          class="navitem navbar-text"
        >
          <span
            class="CartIconBadge position-relative undefined"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-shopping-cart fa-w-18 fa-2x "
              data-icon="shopping-cart"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M528.12 301.319l47.273-208C578.806 78.301 567.391 64 551.99 64H159.208l-9.166-44.81C147.758 8.021 137.93 0 126.529 0H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24h69.883l70.248 343.435C147.325 417.1 136 435.222 136 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-15.674-6.447-29.835-16.824-40h209.647C430.447 426.165 424 440.326 424 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-22.172-12.888-41.332-31.579-50.405l5.517-24.276c3.413-15.018-8.002-29.319-23.403-29.319H218.117l-6.545-32h293.145c11.206 0 20.92-7.754 23.403-18.681z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
      </div>
    </nav>
  </div>
  <div
    style="opacity: 0.0128;"
  >
    <div
      class="my-4 py-4 Home dark-gradient1 rounded px-2 px-xl-5 container-xl"
    >
      <h1
        class="brand-style"
      >
        Widget Store 
      </h1>
      <h2
        class="lead font-italic mb-4 Home-byline"
      >
        We have what you need!
      </h2>
      <div
        class="justify-content-center pb-2 card-deck"
      >
        <div
          class="mx-auto d-inline-block"
        >
          <a
            class="darkLink"
            href="//product/1"
          >
            <div
              class="ProductCard mb-3 card"
            >
              <div
                class="img-wrapper"
              >
                <img
                  alt="Samsung XG-900 55 inch LCD TV"
                  class="card-img"
                  src="https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg"
                />
              </div>
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                  tag="h5"
                >
                  Samsung XG-900 55 inch LCD TV
                </div>
                <div
                  class="mb-2 text-muted card-subtitle h6"
                  tag="h6"
                >
                  <em>
                    Latest generation smart televison 2021
                  </em>
                </div>
                <p
                  class="card-text"
                >
                  <span
                    class="sc-dlfnbm kJMwKE"
                  >
                    <span
                      class="sc-hKgILt gCMyNV price"
                    >
                      $400.40
                    </span>
                  </span>
                </p>
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
